### PR TITLE
fallout-ce: init at 1.0.0

### PR DESCRIPTION
**Pull request by hughobrien**
###### Description of changes

Introduce [fallout-ce](https://github.com/alexbatalov/fallout1-ce), an open source re-implementation of the classic game.

Note this is the engine only, the original game assets are still required (available on digital platforms). Build steps are adapted from their [CI](https://github.com/alexbatalov/fallout1-ce/blob/55fb75bc7799830ac1a6dab71555fb9ef8c1e690/.github/workflows/release.yml#L146).

See also: [fallout2-ce PR](https://github.com/NixOS/nixpkgs/pull/239451)

REVIEW REQUESTS:

1. ~This game uses the 'Sustainable Use License' which I do not see as listed in `lib/licences.nix` - how should we proceed? [Here](https://github.com/NixOS/nixpkgs/blob/a2a7650c472726517c303bdb55d784829cf3ebdf/pkgs/applications/networking/n8n/default.nix#L43) is the only other usage of it within `nixpkgs`.~ Edit: I've made a [PR](https://github.com/NixOS/nixpkgs/pull/239455) for the licence.
1. The binary name is `fallout-ce` but the project is branded `fallout1-ce`, I have hewn to the original name but am open.
1. I was unable to have `substituteInPlace` handle the `${foo}` strings in the Makefile without triggering nix interpolation, so used a two pass process instead. I'm sure there's a better way of doing this.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/games/fallout-ce/default.nix
Comment (SuperSandro2000) [inline pkgs/games/fallout-ce/default.nix:19]: same as fallout2-ce, we can share this between the packages

File: pkgs/games/fallout-ce/default.nix
Comment (SuperSandro2000) [inline pkgs/games/fallout-ce/default.nix:26]: cmake goes to nativeBuildInputs

File: pkgs/games/fallout-ce/default.nix
Comment (SuperSandro2000) [inline pkgs/games/fallout-ce/default.nix:31]:
```suggestion
    substituteInPlace third_party/fpattern/CMakeLists.txt \
      --replace "FetchContent_Populate" "#FetchContent_Populate" \
      --replace "{fpattern_SOURCE_DIR}" "${finalAttr.fpattern}" \
      --replace "$/nix/" "/nix/"
```
please move this to postPatch

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by hughobrien** [comment]
@SuperSandro2000 fpattern PR https://github.com/NixOS/nixpkgs/pull/240569

File: pkgs/games/fallout-ce/default.nix
Comment (hughobrien) [reply pkgs/games/fallout-ce/default.nix]: packaged

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
Doesn't start for me
```
 ➜ ./result/bin/fallout-ce
free(): double free detected in tcache 2
Aborted (core dumped)
```

**Comment by hughobrien** [comment]
I haven't seen that error yet, are you on x64?

Remember the game needs the original assets, and further is particular about the casing of the file names ([they describe this here])https://github.com/arp242/fallout1-ce#configuration)).

I've tested it with the copy I have from Steam, but also from [this copy on the Archive](https://archive.org/details/FalloutAPostNuclearRolePlayingGameUSA).

The correct way is to edit the `fallout.cfg` file but we can just use links as a test:

```bash
unar Fallout%20%28USA%29.zip
cd Fallout%20%28USA%29
unar "Fallout - A Post Nuclear Role Playing Game (USA).bin"
cd "Fallout - A Post Nuclear Role Playing Game (USA)"
ln MASTER.DAT master.dat
ln CRITTER.DAT critter.dat
ln -s DATA data
fallout-ce
```

**Comment by hughobrien** [comment]
How about:

"`fallout-ce` is a community re-implmentation of the Fallout game engine, it does not include the necessary game data to run the game. To use `fallout-ce`, you must first obtain a copy of Fallout (available from e.g. Steam, GoG, or the Internet Archive). Depending on the version you acquire, some files may need to be renamed (in a case sensitive manner), or alternatively the `fallout.cfg` file updated to match. The game will not start if it cannot find the files it expects. By default the files `master.dat`, `critter.dat` and the `data` directory are all expected to be lowercase. For full details refer to the project's installation guide: https://github.com/alexbatalov/fallout1-ce#configuration"

**Comment by NobbZ** [comment]
There are other programs, that require much more setup.

Those do **not** have such a "`longDescription`". In my opinion this should not block the merge.

Also, there is no such thing for eg. `openmw` which also requires original game data.

In my opinion, people who want to use this software usually know in advance that they need the original game data, and if we want to provide helpers for setting everything up, we can provide them in another PR. This details should not block this PR to get the package in.

This extra PR could provide a wrapper and/or a module.

**Comment by peterhoeg** [comment]
I have a couple of comments:

1. openmw will look for its data files in a fixed, known path. If `fallout-ce` doesn't do that, I suggest creating an unwrapped version that used as input for a wrapper that also takes the assets.
2. fallout-ce and fallout2-ce should absolutely be combined. `callPackages` is your friend here. Take a look at `mpd` as an example showing how to do that.

**Comment by peterhoeg** [comment]
Furthermore, I would probably have pulled in `fpattern` locally rather than "polluting" the global namespace as it's local to these 2.

**Comment by hughobrien** [comment]
> Furthermore, I would probably have pulled in `fpattern` locally rather than "polluting" the global namespace as it's local to these 2.

There's [some disagreement](https://github.com/NixOS/nixpkgs/pull/239450#discussion_r1246814734) on that, but I'm happy to remove if there's consensus

**Comment by TheBrainScrambler** [comment]
I had tried myself to package fallout-ce, turns out it was useless.
I would personally suggest naming it fallout1-ce, because it's more consistent with fallout2-ce. I must say I had renamed the binary to fallout1-ce too.

For reference my stuff, but yours is better anyway:
https://0x0.st/H_tq.nix fallout1-ce.nix
https://0x0.st/H_tb.patch cmake-dont-fetch-deps.patch

In my case I had only considered fallout1-ce, but it's probably better to package fpattern as you did because that way it is built only once and not twice.

> I have a couple of comments:
> 
>     1. openmw will look for its data files in a fixed, known path. If `fallout-ce` doesn't do that, I suggest creating an unwrapped version that used as input for a wrapper that also takes the assets.
fallout1-ce only looks for data in the current directory. So yes I would also go for a wrapper, though I would make a simple one that just changes the current directory to ~/.local/share/fallout1-ce.

>     2. fallout-ce and fallout2-ce should absolutely be combined. `callPackages` is your friend here. Take a look at `mpd` as an example showing how to do that.
Totally agree here. I would like to do that myself if I can. Would need to do a PR for the PR ;) I wonder how to do that.

> Furthermore, I would probably have pulled in `fpattern` locally rather than "polluting" the global namespace as it's local to these 2.
Maybe some other project that hasn't be packaged yet uses it ? I would rather say remove it from the global namespace if a conflict ever occurs. In that case though I could still keep fpattern as a package, but as one that only fallout1-ce and fallout2-ce would see and that wouldn't be exported, meaning it wouldn't be declared in top-level/all-packages.nix

**Comment by TheBrainScrambler** [comment]
> There are other programs, that require much more setup.
> 
> Those do **not** have such a "`longDescription`". In my opinion this should not block the merge.
> 
> Also, there is no such thing for eg. `openmw` which also requires original game data.
openmw will ask you for the location of the assets if you don't provide those, with nice GUI menus and all, so it's not a problem if you don't explain how to install assets because openmw will tell you. fallout1-ce on the other hand will just crash with a cryptic error, so here an explanation would be needed.

> In my opinion, people who want to use this software usually know in advance that they need the original game data, and if we want to provide helpers for setting everything up, we can provide them in another PR. This details should not block this PR to get the package in.
I agree and disagree. It maybe shouldn't block the PR, but it would be a good idea to give an explanation. I disagree that people usually know in advance that they need the assets because when you just search for games in the nixpkgs, games that you don't know, then you don't know if they need assets or not.

> This extra PR could provide a wrapper and/or a module.
As said above I would go for something simple that cd into ~/.local/share/fallout1-ce and gives an explanation telling you to put assets there. For people who don't want that there would still be the unwrapped version.

**Comment by hughobrien** [comment]
These have been growing stale so let's review.

Breakout common `fpattern` dependency
-------------------------------------------------------
* [Requested here](https://github.com/NixOS/nixpkgs/pull/239450#discussion_r1246662816)
* [Implemented here](https://github.com/NixOS/nixpkgs/pull/240569)
* [Challenged here](https://github.com/NixOS/nixpkgs/pull/239450#issuecomment-1647327751)
* [Defended here](https://github.com/NixOS/nixpkgs/pull/239450#issuecomment-1673757779)

I doubt it will be used beyond these two titles as it is odd code, however it is already packaged so reverting seems counter-productive. Is this a merge blocker?

Rename fallout-ce to fallout1-ce
-------------------------------------------
* [Requested here](https://github.com/NixOS/nixpkgs/pull/239450#issuecomment-1673757779)

I am against this as it diverges from upstream, indeed the upstream repos are named `fallout1-ce` and `fallout2-ce`, but the binaries within are named `fallout-ce` and `fallout2-ce`. This is in line with their original release. The game is named [Fallout](https://www.imdb.com/title/tt0134648/), not Fallout 1. `nixpkgs` does not call Doom `doom1`.

Merge into a single derivation
----------------------------------------
* Requested [here](https://github.com/NixOS/nixpkgs/pull/239450#issuecomment-1647325155) and [here](https://github.com/NixOS/nixpkgs/pull/239450#issuecomment-1673757779)

Note: I am assuming, based on the `mpd` suggestion that a single derivation was what was meant.

I am strongly against merging 1 & 2 into a single derivation.
1. They are separate upstream, we do not know how they intend to diverge
3. They are already slightly different (zlib dependency)
4. The games themselves are entirely different SKUs
5. `mpd` as the example, still only pulls from one repo, the fallouts are different repos.

I suggest instead that we create a `pkgs/games/fallout` directory with the two derivations inside, similar to `quake3`.

```
$ ls nixpkgs/pkgs/games/quake3/
content  ioquake  quake3e  wrapper
```

More arguments against a single derivation:
* Quake 2 and 3 are different derivations. Perhaps one could argue the underlying engines are different enough to warrant this.
* The `alephone` implementations of Marathon and Marathon-Durandal, are different derivations, but are in fact a unified engine for [multiple games](https://alephone.lhowon.org/scenarios.html)

The game requires some setup
------------------------------------------
Mentioned several times. These games, like several others in `nixpkgs` require external non-distributable assets.

Upstream's intention is for the binary to be a drop-in replacement for the original binary in the game's installation folder. The [upstream instructions](https://github.com/alexbatalov/fallout1-ce#linux) are quite clear, 'install' the game through Wine or similar and drop in the replacement binary. It will search the working directory for the game assets. It's not clear to me that we should be prescriptive as to where the user places these assets, though the `~/.local` suggestion is sensible.

A further complication that upstream addresses is that different releases of the game will have different asset names, made more troublesome on a modern FS due to FAT's case-insensitivity. 

Upstream's guidance is to either rename the files to be as expected or to set the paths in the game's config file.

The question before us seems to be: How much should we smooth out the experience of using this package, beyond what upstream has done?

I think the only reasonable answer is "as little as possible" since such work ultimately should be a PR to upstream. However the silent failure is definitely something we can improve.

I suggest instead a simple wrapper that checks for the default asset files (perhaps in the PWD and a defined `~/.local` dir) and exits with a helpful message and link to the official documentation if not found ([suggested here](https://github.com/NixOS/nixpkgs/pull/239450#issuecomment-1630801264)).

I suggest that such a wrapper follow in a future PR as there is clearly interest in this as-is.

Improve the long-description
---------------------------------------
No objections, some suggestions given in the thread.


Apologies if I missed any feedback here, please let met know.

**Comment by TheBrainScrambler** [comment]
> ## Rename fallout-ce to fallout1-ce
> 
>     * [Requested here](https://github.com/NixOS/nixpkgs/pull/239450#issuecomment-1673757779)
> 
> 
> I am against this as it diverges from upstream, indeed the upstream repos are named `fallout1-ce` and `fallout2-ce`, but the binaries within are named `fallout-ce` and `fallout2-ce`. This is in line with their original release. The game is named [Fallout](https://www.imdb.com/title/tt0134648/), not Fallout 1. `nixpkgs` does not call Doom `doom1`.
I agree, you convinced me on this.

> ## Merge into a single derivation
> 
>     * Requested [here](https://github.com/NixOS/nixpkgs/pull/239450#issuecomment-1647325155) and [here](https://github.com/NixOS/nixpkgs/pull/239450#issuecomment-1673757779)
> 
> 
> Note: I am assuming, based on the `mpd` suggestion that a single derivation was what was meant.
Ah I misunderstood that suggestion. What I suggested was not to make a single derivation. Because the build process is very similar, I suggested to move all the generic steps in a function in say `generic.nix` and it would then be called in `default.nix` for each fallout-ce version. For example like `sm64ex` and `fteqw` are doing it. In the case of `sm64ex` it handles 2 differents upstreams, but it doesn't matter since the build steps are similar. This would have avoided SuperSandro to review twice almost the same nix files ;)

> ## The game requires some setup
> 
> Mentioned several times. These games, like several others in `nixpkgs` require external non-distributable assets.
> 
> Upstream's intention is for the binary to be a drop-in replacement for the original binary in the game's installation folder. The [upstream instructions](https://github.com/alexbatalov/fallout1-ce#linux) are quite clear, 'install' the game through Wine or similar and drop in the replacement binary. It will search the working directory for the game assets. It's not clear to me that we should be prescriptive as to where the user places these assets, though the `~/.local` suggestion is sensible.
> 
> A further complication that upstream addresses is that different releases of the game will have different asset names, made more troublesome on a modern FS due to FAT's case-insensitivity.
> 
> Upstream's guidance is to either rename the files to be as expected or to set the paths in the game's config file.
> 
> The question before us seems to be: How much should we smooth out the experience of using this package, beyond what upstream has done?
> 
> I think the only reasonable answer is "as little as possible" since such work ultimately should be a PR to upstream. However the silent failure is definitely something we can improve.
> 
> I suggest instead a simple wrapper that checks for the default asset files (perhaps in the PWD and a defined `~/.local` dir) and exits with a helpful message and link to the official documentation if not found ([suggested here](https://github.com/NixOS/nixpkgs/pull/239450#issuecomment-1630801264)).
I suppose the right thing to do is that upstream implements that case insensitivity ?

I would also suggest putting the helpful message in longDescription, as you proposed previously.

And again we can just make 2 packages, one unwrapped and one with wrapper, to make everyone happy.

**Comment by TheBrainScrambler** [comment]
I've just implemented my suggestion with `generic.nix`.
Not sure how I can use github to share it so it is at https://0x0.st/H_lF.tar.gz
What do you think ?

**Comment by hughobrien** [comment]
@TheBrainScrambler you should be able to PR to my branch https://github.com/hughobrien/nixpkgs/tree/fallout-ce - then it'll get merged together. I took at look at your files and I like it, thank you.

A simple wrapper that prevents @SuperSandro2000 's experience with some basic guidance seems pretty reasonable to me. I confess, for my particular assets (from Steam) that I had to `strace` to see what the binary was looking for and there's no need to propagate that pain.

**Comment by hughobrien** [comment]
Are there strong feelings against pulling down one of archive.org's zips as part of the build process? I can find only one example of this in `nixpkgs` [here](https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/games/wargus/default.nix)

**Comment by hughobrien** [comment]
@TheBrainScrambler I've implemented your changes, would you mind taking a look? Also added you as a maintainer, hope you don't think it presumptive.

**Comment by TheBrainScrambler** [comment]
> Are there strong feelings against pulling down one of archive.org's zips as part of the build process? I can find only one example of this in `nixpkgs` [here](https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/games/wargus/default.nix)

Yes there are: https://github.com/NixOS/nixpkgs/issues/208078
I believe fetching the full game from archive.org, or anywhere else, would be illegal.
From what I've found this asset downloading was committed in https://github.com/NixOS/nixpkgs/pull/185319 for an update of the package, which made it a bit hidden I guess ?
Now I don't know if providing a script to download the assets from somewhere illegal is permitted ? In any case the assets would end up as a different package since you can get them from many sources, be it from Steam, archive.org or anywhere else.

My take on this again would be to also provide a wrapper that chdir into ~/.local/share/fallout-ce and documents this somewhere, because that's how many other programs do it: https://github.com/NixOS/nixpkgs/issues/208078#issuecomment-1366920345, I can also name OpenXcom, doom and quake ports, the Force Engine, gemrb, etc ... They normally all expect you to put the assets into some directory and read those at runtime.
Wargus was apparently an exception because it needs the assets at compile-time. sm64ex would be another such example.

**Comment by TheBrainScrambler** [comment]
> @TheBrainScrambler I've implemented your changes, would you mind taking a look? Also added you as a maintainer, hope you don't think it presumptive.

No it's fine to have me as maintainer, I would have eventually created a PR for this too had you not done it.
Did a review as PR: https://github.com/hughobrien/nixpkgs/pull/1

**Comment by hughobrien** [comment]
I used `--run cd X` instead of `--chdir` in makeWrapper as `chdir` doesn't appear to be able to handle dynamic targets (such as `$HOME`). Alternatives suggestions welcome.

**Comment by TheBrainScrambler** [comment]
> I used `--run cd X` instead of `--chdir` in makeWrapper as `chdir` doesn't appear to be able to handle dynamic targets (such as `$HOME`). Alternatives suggestions welcome.

Looks like there is the same problem at https://github.com/NixOS/nixpkgs/pull/248915/files (default.nix line 54). It got approved so I suppose right now there is no better way ...

**Comment by TheBrainScrambler** [comment]
I also noticed a problem: the wrapper ends up in `$out/bin/wrapper.sh`, but that means it will end up polluting $PATH. A simple workaround could be to rename it to `.wrapper.sh`. But it could be nice to use writeShellScript instead. I just fear that there might be a problem at line 7 of `wrapper.sh` with "${REQUIRED_FILES[@]}", which might be expanded by nix ? I suck at bash so would need your advice here.

**Comment by hughobrien** [comment]
Good catches, fixed both I think by cribbing from [Rise of the Triad](https://github.com/NixOS/nixpkgs/blob/nixos-23.05/pkgs/games/rott/default.nix#L14). 

unwrapped binary goes to libexec and we do the `cd` in the wrapper along with a direct `exec`.

**Comment by TheBrainScrambler** [comment]
https://github.com/hughobrien/nixpkgs/pull/2

**Review by TheBrainScrambler — APPROVED** [review]

**Comment by TheBrainScrambler** [comment]
I just forgot something: you should squash all these commits into a single one with message "fallout-ce: init at 1.0.0". Though this should also include fallout2-ce somehow ... Not sure how it should be done, but there is definitively a need for squashing some commits.

**Comment by hughobrien** [comment]
Yes absolutely, I'm just waiting for other feedback so I can avoid doing it twice.

**Review by TheBrainScrambler — APPROVED** [review]

**Comment by TheBrainScrambler** [comment]
Well I was re-requested for a review because of the rebasing, so I suppose you didn't loose your time, but instead avoided the need for 2 reviews from others :p

File: pkgs/games/fallout-ce/build.nix
Comment (peterhoeg) [inline pkgs/games/fallout-ce/build.nix:65]: A minor thing - we don't want the unwrapped binary (assuming that it doesn't work unless wrapped) to "pollute" the global PATH, so better stick it in $out/libexec and adjust the wrapper.

Comment (TheBrainScrambler) [reply pkgs/games/fallout-ce/build.nix:65]: Lol this is exactly what was done before, and I insisted on getting the unwrapped version. The unwrapped binary still works, it's just that it will try to load the assets in the current directory only, while the wrapper will use $XDG_DATA_HOME. I wanted to keep that unwrapped binary in case someone doesn't want assets in $XDG_DATA_HOME and instead wants to make his own wrapper (or no wrapper at all).

Comment (hughobrien) [reply pkgs/games/fallout-ce/build.nix:65]: I don't mind either way, there are a few examples of the `libexec` approach already as this assets dir problem is common for games.

1. [rott](https://github.com/NixOS/nixpkgs/blob/master/pkgs/games/rott/default.nix#L52)
2. [vessel](https://github.com/NixOS/nixpkgs/blob/master/pkgs/games/vessel/default.nix#L64)
3. [openmw](https://github.com/NixOS/nixpkgs/blob/master/pkgs/games/openmw/tes3mp.nix#L139)
4. [nethack](https://github.com/NixOS/nixpkgs/blob/master/pkgs/games/nethack/default.nix#L143)

@TheBrainScrambler Unless you personally have a use case for the `-unwrapped` being on PATH then I'm leaning more towards YAGNI.

Comment (TheBrainScrambler) [reply pkgs/games/fallout-ce/build.nix:65]: I don't have an use case myself. It's interesting, I thought there would be more packages exposing their unwrapped version, it's good to know that I was wrong. There's just one lol:
- [0ad](https://github.com/NixOS/nixpkgs/blob/master/pkgs/games/0ad/default.nix#L6)

There are also packages that somehow provide an unwrapped version but which won't run because they are missing libraries otherwise ?
- [prismlauncher](https://github.com/NixOS/nixpkgs/blob/6e1fe01570b8eed68e5969ee6c5fc77159891ed0/pkgs/top-level/all-packages.nix#L38014)
- [xonotic](https://github.com/NixOS/nixpkgs/blob/6e1fe01570b8eed68e5969ee6c5fc77159891ed0/pkgs/top-level/all-packages.nix#L38425)

So perhaps the right approach would be instead to not expose the unwrapped version, and fix this in those packages that I listed too.

Comment (hughobrien) [reply pkgs/games/fallout-ce/build.nix]: @peterhoeg moved to libexec

**Review by TheBrainScrambler — APPROVED** [review]

**Comment by hughobrien** [comment]
Polite request for a merge if everyone is satisfied.

**Comment by hughobrien** [comment]
Thank you @amaxine ! Please allow me to push my luck and invite you to take a look at [this dosbox-x PR](https://github.com/NixOS/nixpkgs/pull/258301)

**Pull request merged**
